### DISCUSSION TITLE
Change segmentation directory structure

### DISCUSSION THREAD

**Discussion by ghost** [Project-MONAI/MONAI#2069] [Q&A]
Hi,
I'm trying to save my completed segmentations to a different directory structure than default: rather than folders of name 0,1,2.. I'd like to use the input case/image names: S01,S02,S03...

The code I'm using is:
```
saver = NiftiSaver(output_dir=outputdata_dir, dtype=(['numpy.float16']))
for i, val_data in enumerate(test_loader):
    [other stuff]
    saver.save_batch(largest)
```

I realise there is now SegmentationSaver, but the project is wrapping up so I'd ideally get it working with my current implementation, if possible? I've tried reading the docs and spent ages trying to adjust it myself but with no luck.

Alternatively it is is easier I'd be happy to just save all the segmentations to the main output_dir but to change the output filename to each case name eg S01_seg, S02_seg..., but again I can't figure out how to modify the default 0,1,2 etc

Any assistance would be much appreciated!

Edit: To be clear, I know in principal how to set things to the case/image ID (eg f"{case_name[i]}") but my issue is how/where I can add this arguement.

**Comment by ghost** [comment]
I'm returning to this problem and any help would be much appreciated please. Thanks in advance

**Comment by Nic-Ma** [comment]
Hi @jnc13 ,

Thanks for your interest here.
I think maybe you missed to pass the `meta data` to the saver, so it uses index as file names:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/data/nifti_saver.py#L118
Let me try to update the doc-string to make it more clear.

Thanks.

**Reply by Nic-Ma** [reply]
The document had been enhanced.

Thanks.

**Reply by Meddebma** [reply]
Hi @Nic-Ma 

I have the same problem reading the meta data of my files, I wrote:
saver.save(val_outputs, meta_data="filename_or_obj") 

did I write it wrong?

here is my code
https://github.com/Meddebma/pyradiomics/blob/master/3D_Segmentation_eval.ipynb

thanks

**Reply by Nic-Ma** [reply]
Hi @Meddebma ,

I think you should use `saver.save_batch(val_outputs, val_data["image_meta_dict"])` instead.
Please refer to this example:
https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/torch/unet_evaluation_dict.py#L91.

Thanks.

**Reply by Meddebma** [reply]
@Nic-Ma,

I tried it like that, I've got this error:
TypeError: expected str, bytes or os.PathLike object, not list
Here is the code:
https://github.com/Meddebma/pyradiomics/blob/master/3D_Segmentation_eval.ipynb

thanks

**Reply by Nic-Ma** [reply]
I think that because your file path contains blank:
`ima_dir= "/content/drive/My Drive/Spleen_AI/Projekt2ß"`
Maybe you can change to some regular path and try again?

Thanks.

**Reply by Meddebma** [reply]
I removed the blank (MyDrive works as well) and I still got the same issue

https://github.com/Meddebma/pyradiomics/blob/master/3D_Segmentation_eval.ipynb
thanks

**Reply by Nic-Ma** [reply]
I didn't get what's wrong here.. could you please help print out `val_data["image_meta_dict"]` before saving?

Thanks.

**Reply by Meddebma** [reply]
Here is what I get when I print it out:
https://github.com/Meddebma/Complete-Python-3-Bootcamp/blob/master/3D_Segmentation_eval.ipynb

**Reply by Nic-Ma** [reply]
I got your root cause: please use `saver.save_batch()` as the example I provided in the previous comment. You used `saver.save()` I think.

Thanks.

**Reply by Meddebma** [reply]
It worked @Nic-Ma  :)
thank you very much!

**Reply by Nic-Ma** [reply]
You are welcome.
Please feel free to raise a ticket if you have any more questions.

Thanks.

**Reply by Meddebma** [reply]
@Nic-Ma thank you!
I exported the files and opened them in a Nifti Reader, unfortunately the segmentation doesnt match the coordinate of  the volume, should I try to put Invertd here?

**Reply by Nic-Ma** [reply]
Yes, I think you can play with `Invertd` to make sure the prediction matches input image. Refer to this example carefully:
https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/torch/unet_inference_dict.py

Thanks.

**Reply by Meddebma** [reply]
@Nic-Ma 
when I try this example with pip install monai-weekly some of the dependencies dont work, which version should I install?

**Reply by Nic-Ma** [reply]
I would suggest the latest monai-weekly.
You can check the `requirement-dev.txt` for the dependency.

Thanks.

**Reply by ghost** [reply]
> 
> 
> Hi @jnc13 ,
> 
> Thanks for your interest here.
> I think maybe you missed to pass the `meta data` to the saver, so it uses index as file names:
> https://github.com/Project-MONAI/MONAI/blob/dev/monai/data/nifti_saver.py#L118
> Let me try to update the doc-string to make it more clear.
> 
> Thanks.

Hi Nic,

Thanks for updating the document, though I am still a little confused about the implementation.

Currently I can only get it to take the first character of the 'filename_or_obj' dictionary entry from the meta_data argument. 

Concretely if I for example were to use this simplified example:

```
case_name_test = {'filename_or_obj': 'TEST'}
saver.save_batch(val_output_bin, meta_data = case_name_test)
```

I output a segmentation with directory & name: "....T\T_seg.nii.gz."

My cases are labelled S01,S02,S03 etc so ideally I want to be able to use the whole case name as the filename. Could you please share how this is done?

Related to this question, is there a way to forfeit saving the segmentations each into their own folder? I've read the docstring and followed the example is the "data_root_dir" argument but can't see a way, in that example, to avoid adding the "/image/" structure?

Thank you!

**Reply by Nic-Ma** [reply]
Hi @jnc13 ,

I don't quite understand your question, I think if the input filename is "TEST", the output will be:
```py
from monai.data import create_file_basename

print(create_file_basename(postfix="seg", input_file_name="TEST", folder_path="/output"))
>>>/output/TEST/TEST_seg
```
Could you please help share more details about your question?
@rijobro Maybe I misunderstood something here?

Thanks.

**Reply by ghost** [reply]
Hi @Nic-Ma,

I think I must be implementing the meta_data argument incorrectly, could you please re-explain how to have it take the input filename?

For a separate purpose as part of my script I create a list of the filenames and I tried passing entries of this to the dictionary but had the above problem where it was only taking the first letter of the filename, not the whole thing. I'm likely missing a way to directly take the filename as the "filename_or_obj" dictionary entry?

Thanks,

**Reply by Nic-Ma** [reply]
Hi @jnc13 ,

I see your problem.
If you call the `save_batch()` function of saver, it expects the meta_data is collated by DataLoader, so the `filename_or_obj` should be a **list** for 1 batch:
```
meta_dict = {'filename_or_obj': ['TEST', ...]}
```
If you only save for 1 image, please call `saver.save()` instead of `save_batch()`.

Thanks.

**Reply by ghost** [reply]
Thanks @Nic-Ma. So I am trying to save multiple images and am using `save_batch()`.

I've tried again with a slightly amended approach and managed to almost get it to now work (yay!) so thank you.
Separately for my script I have a list `case_name`:

`List of cases:  ['S01', 'S05']`

I then fed this as in your example:

`saver.save_batch(val_output_bin, meta_data = {'filename_or_obj': case_name})`

And it almost works but runs into a problem because I have outputs:

>/output/S01/S01_seg  --- correct
>/output/S05/**S01**_seg --- not correct

The S05 case filename is mislabelled S01, how do I avoid this? If there is a more direct implementation, using the meta_data collated by the DataLoader as you mention then I would be keen to hear it if it can help avoid this issue.

Thanks

**Reply by Nic-Ma** [reply]
Hi @jnc13 ,

Could you please help provide a simple tiny program to reproduce your issue?
I really can't figure out why it generates `/output/S05/S01_seg`. Maybe we have a bug somewhere.

Thanks in advance.

**Reply by ghost** [reply]
Hi @Nic-Ma,

Beyond pasting in my whole script and provided some of the images I'm not actually sure how best to cut it down to a tiny usable script since so much of the code is required to get it to run without an error.

So I have a small function to extract the case name and save it to a list `case_name` as shown above, and then for outputting segmentations I have:
```

model.load_state_dict(torch.load(trained_model))
model.eval()
with torch.no_grad():
     for image in range(len(case_name)):
            print(f'Processing {case_name[image]}')
            val_input = test_ds[image]["image"].unsqueeze(0).to(device)
            post_trans = Compose([Activations(sigmoid=True), AsDiscrete(threshold_values=True, logit_thresh=0.7)])  
    
            val_output = model(val_input)
            val_output_bin = post_trans(val_output)

            saver = NiftiSaver(output_dir=outputdata_dir+ f'/{case_name[image]}', dtype=(['numpy.float16']))

            saver.save_batch(val_output_bin, meta_data = {'filename_or_obj': case_name})
            print(val_output_bin.shape)
```

I realise I actually slightly miscommunicated the output, there is an additional directory layer. When I run this I get:

/output/S01/S01/S01_seg
/output/S05/S01/S01_seg

I think the only correct bit of this structure actually stems from the `/{case_name[image]}` portion of `saver = NiftiSaver(output_dir=outputdata_dir+ f'/{case_name[image]}', dtype=(['numpy.float16']))`.

If instead I run it without `/{case_name[image]}` as in: `saver = NiftiSaver(output_dir=outputdata_dir, dtype=(['numpy.float16']))`

It does not correctly loop over the cases in `case_name`:
>Processing S01
>file written: C:\Users\...\output\S01\S01_seg.nii.gz.
>torch.Size([1, 2, 64, 64, 64])
>Processing S05
>file written: C:\Users\...\output\S01\S01_seg.nii.gz.
>torch.Size([1, 2, 64, 64, 64])

Which of course only results in one file rather than two:
/output/S01/S01_seg,nii.gz

Any ideas appreciated. 

Thanks.

**Reply by ghost** [reply]
Since it is a related question to this discussion, is there a way to avoid having a new directory being created for each segmentation being saved? Ideally I would have the segmentation files just saved to the output directory rather than in separate folders:

/output/S01_seg.nii.gz
/output/S05_seg.nii.gz
etc

Thanks

**Reply by Nic-Ma** [reply]
Hi @jnc13 ,

Thanks for your feedback.
Let me try to enhance the saver to make it possible to not save them in separate folders.

Thanks.

**Reply by ghost** [reply]
That would be great, thanks @Nic-Ma.
I'm still a little stuck getting the saver output filenname to be the input filename. I think maybe rather than using my own list I should try to do as you suggested: "If you call the `save_batch()` function of saver, it expects the meta_data is collated by DataLoader, so the filename_or_obj should be a list for 1 batch:

`meta_dict = {'filename_or_obj': ['TEST', ...]}`
"
Could you please clarify how I can go about accessing the meta_data collated by the DataLoader to use as the list for meta_dict?

Thanks as always, I really do appreciate your help!

**Reply by Nic-Ma** [reply]
Hi @jnc13 ,

I think you can check this example deeply to see how it works: https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/torch/unet_inference_dict.py
Especially how the `SaveImaged` transform works.

Thanks.

**Reply by Nic-Ma** [reply]
BTW, I created a ticket to track your feature request: https://github.com/Project-MONAI/MONAI/issues/2562.

Thanks.

**Reply by Nic-Ma** [reply]
Hi @jnc13 ,

Submitted PR to add support for your feature request: https://github.com/Project-MONAI/MONAI/pull/2576.

Thanks.
